### PR TITLE
[Merged by Bors] - feat(FieldTheory.Separable) : generalize `IsSeparable.of_equiv_equiv` to noncommutative cases

### PR DESCRIPTION
**Pull request by Yu-Misaka**
Greetings! In this PR I attempt to generalize `IsSeparable.of_equiv_equiv` and some related codes to noncommutative cases. This could be useful in #16525 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Yu-Misaka** [comment]
Some major problems unsolved:

1. I want `algebraComp` in Mathlib/Algebra/Algebra/Defs.lean (line 366) to be an instance instead of definition, so that I can get rid of all `haveI` in the statement.
2. Regarding the two `sorry` in Mathlib/FieldTheory/Separable.lean (line 732), the annotated `exact` gives exactly the same things but they don't seem to work.

**Comment by Yu-Misaka** [comment]
Right. The first problem seems impossible to solve. And I think ```IsSeparable.of_equiv``` and ```algebraComp``` is more or less useless and confusing, so I removed them. Then there shouldn't be any other problems.

File: Mathlib/FieldTheory/Separable.lean [hunk 1]
```diff
@@ -679,25 +679,29 @@ end Field
...
 
 lemma IsSeparable.of_equiv_equiv {x : B₁} (h : IsSeparable A₁ x) : IsSeparable A₂ (e₂ x) :=
   letI := e₁.toRingHom.toAlgebra
-  letI := ((algebraMap A₁ B₁).comp e₁.symm.toRingHom).toAlgebra
-  haveI : IsScalarTower A₁ A₂ B₁ := IsScalarTower.of_algebraMap_eq
-    (fun x ↦ by simp [RingHom.algebraMap_toAlgebra])
+  letI : Algebra A₂ B₁ :=
+    { (algebraMap A₁ B₁).comp e₁.symm.toRingHom with
+        smul := fun a b ↦ ((algebraMap A₁ B₁).comp e₁.symm.toRingHom a) * b
+        commutes' := fun r x ↦ (alg1.commutes) (e₁.symm.toRingHom r) x
```
Comment (jjdishere) [inline Mathlib/FieldTheory/Separable.lean:694]:
```suggestion
        commutes' := fun r x ↦ (Algebra.commutes) (e₁.symm.toRingHom r) x
```

File: Mathlib/FieldTheory/Separable.lean [hunk 2]
```diff
@@ -679,25 +679,29 @@ end Field
 
 section AlgEquiv
 
-variable {A₁ B₁ A₂ B₂ : Type*} [Field A₁] [Field B₁]
-    [Field A₂] [Field B₂] [Algebra A₁ B₁] [Algebra A₂ B₂] (e₁ : A₁ ≃+* A₂) (e₂ : B₁ ≃+* B₂)
+open RingHom RingEquiv
+
+variable {A₁ B₁ A₂ B₂ : Type*} [Field A₁] [Ring B₁] [Field A₂] [Ring B₂]
+    [alg1 : Algebra A₁ B₁] [Algebra A₂ B₂] (e₁ : A₁ ≃+* A₂) (e₂ : B₁ ≃+* B₂)
```
Comment (jjdishere) [inline Mathlib/FieldTheory/Separable.lean:685]: I think it is possible to omit the instance name `alg1` here.

**Review by jjdishere — COMMENTED** [review]
Thanks! Except for a minor suggestion, all changes seem good to me!

File: Mathlib/FieldTheory/Separable.lean (hunk 1 above)
Comment (Yu-Misaka) [reply Mathlib/FieldTheory/Separable.lean:694]: Thanks! Implemented.

**Review by YaelDillies — APPROVED** [review]
Hey! Sorry for the huge delay. This PR looks good, thank you for submitting 😄 

maintainer merge

**Comment by riccardobrasca** [comment]
!bench

**Comment by riccardobrasca** [comment]
This bench results are a bit weird. I am merging master and benching again.

**Comment by YaelDillies** [comment]
These bench results are reminiscent of @Kha's observation that the times kept on changing by +- several percents due to two OS versions being used simultaneously on the benchmark machines. It will indeed probably disappear after merging master.

**Comment by riccardobrasca** [comment]
!bench

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**